### PR TITLE
Decimal binary operation constants fix

### LINKED ISSUE #20283: wrong toDecimalX minus result
I've been studying basic usage and I found wrong minus result on DecimalXX type.

**Example Query**
`
select toDecimal64(4,4) - toDecimal32(2,2)
`

**Expected output**
2.0000

**Actual output**
3.9800

Is there anything wrong or maybe my misunderstanding of toDecimal ?

image

**How to reproduce**
1. OS: macOS Catalina 10.15.5
2. server: clickhouse-server, built from source, version 21.3.1.1
3. client:  clickhouse-client,  built from source, version 21.3.1.1

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Fixes #20283.

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix incorrect result of binary operations between two constant decimals of different scale. Fixes #20283.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

File: src/Functions/FunctionBinaryArithmetic.h [hunk 1]
```diff
@@ -894,9 +894,8 @@ class FunctionBinaryArithmetic : public IFunction
...
 
             const ResultType res = check_decimal_overflow
-                // the arguments are already scaled after conversion
-                ? OpImplCheck::template process<left_is_decimal, right_is_decimal>(const_a, const_b, 1, 1)
-                : OpImpl::template process<left_is_decimal, right_is_decimal>(const_a, const_b, 1, 1);
+                ? OpImplCheck::template process<left_is_decimal, right_is_decimal>(const_a, const_b, scale_a, scale_b)
+                : OpImpl::template process<left_is_decimal, right_is_decimal>(const_a, const_b, scale_a, scale_b);
 
             if constexpr (result_is_decimal)
                 return ResultDataType(type.getPrecision(), type.getScale()).createColumnConst(
```
Comment (myrrc) [inline src/Functions/FunctionBinaryArithmetic.h:901]: Not sure whether it's the right fix or not, while developing the templated solution I've encountered various bugs due to the fact the `const_float * const_decimal` and vice verca __are__ scaled in `helperGetOrConvert`, so when the `scale_a` and `scale_b` weren't 1, the `process` function produced over- and under-scaling.

The general problem here is that for different operations (multiply, divide, etc...) the process function works differenly (and I couldn't change it while refactoring the initial code.)

Let's wait for the tests.

File: tests/queries/0_stateless/01711_decimal_multiplication.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+SELECT materialize(toDecimal64(4,4)) - materialize(toDecimal32(2,2));
+SELECT toDecimal64(4,4) - materialize(toDecimal32(2,2));
+SELECT materialize(toDecimal64(4,4)) - toDecimal32(2,2);
+SELECT toDecimal64(4,4) - toDecimal32(2,2);
```
Comment (myrrc) [inline tests/queries/0_stateless/01711_decimal_multiplication.sql:4]: My 50 cents: the test is called decimal_multiplication, but the test doesn't contain the multiplication tests.

**Review by myrrc — CHANGES_REQUESTED** [review]

**Comment by alesapin** [comment]
This fix is required for releases. Tests ok, `Thread leak failure` is known https://github.com/ClickHouse/ClickHouse/issues/20032.

File: src/Functions/FunctionBinaryArithmetic.h (hunk 1 above)
Comment (chenbo1989) [reply src/Functions/FunctionBinaryArithmetic.h:901]: As far as I tried, it is the right fix. Thanks

**Pull request merged**